src/app/clusters/fan-control-server/fan-control-server.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/****************************************************************************
+ * @file
+ * @brief Implementation for the Fan Control Server Cluster
+ ***************************************************************************/
+
+#include <math.h>
+
+#include <app-common/zap-generated/af-structs.h>
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::FanControl;
+using namespace chip::app::Clusters::FanControl::Attributes;
+
+// =============================================================================
+// Pre-change callbacks for cluster attributes
+// =============================================================================
+
+using Status = Protocols::InteractionModel::Status;
+
+Protocols::InteractionModel::Status
+MatterFanControlClusterServerPreAttributeChangedCallback(const ConcreteAttributePath & attributePath,
+                                                         EmberAfAttributeType attributeType, uint16_t size, uint8_t * value)
+{
+    Protocols::InteractionModel::Status res;
+
+    switch (attributePath.mAttributeId)
+    {
+    case SpeedSetting::Id: {
+        uint8_t speedMax;
+        EmberAfStatus status = SpeedMax::Get(attributePath.mEndpointId, &speedMax);
+        VerifyOrReturnError(EMBER_ZCL_STATUS_SUCCESS == status, Status::InvalidValue);
+
+        if (*value <= speedMax)
+        {
+            res = Status::Success;
+        }
+        else
+        {
+            res = Status::InvalidValue;
```

Comment:
```suggestion
            res = Status::ConstraintError;
```
InvalidValue is a deprecated name we are trying to remove.